### ISSUE TITLE
Allow supervisors to access specific patient reports so that they can review them

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#4591]
For the Muso IN, we are planning to deploy a workflow where a supervisor needs to review all patient assessments that included an mRDT test. The CHW will be taking a photo of the mRDT test to upload with the patient assessment and the supervisor needs to review the photo to check whether or not the CHW read the result of the test correctly. Since the Muso supervisors are restricted users, they do not have access to all people and all reports about those people. This ticket would allow the supervisor to access reports about patients that meet certain conditions. In this case, it would be patient assessment forms in which an mRDT test was administered.

Requirements:
- Supervisors have to be able to see the photos and the result that the CHW reported after reading the test within a patient assessment form (there are multiple patient assessment forms, one for children and one for adults)
- Supervisors need to be able to use the feature built in #4529 to review the patient assessment form and mark if the result was read correctly or incorrectly
- The CHW needs to know if their result was read correctly or incorrectly because they need to see a target widget that displays their error rate (number or percent of reports marked incorrect/has errors)
- If possible, the supervisor would like to see the error rate for each CHW on that CHW's profile in the supervisor's view of the app

Acceptance criteria:
- When a CHW submits a report with an mRDT photo, the required information successfully replicates to the supervisor's tablet
- After the supervisor marks the report as correct or incorrect, the CHW's error rate target widget updates correctly and the CHW's profile card (as seen by the manager) updates to display the latest error rate

@garethbowen let me know if I've omitted anything from our discussion.

**Comment by garethbowen** [comment]
LGTM

**Comment by SCdF** [comment]
The different ways of doing this that I can see:
 - Idea 0: keep discussing which bits of this we do and do not need to do to refine what we need to build (AFAIK we haven't finished this yet!)
 - Idea 1: add support for server-side task generation and writing tasks to disk, along with support for dynamically working out what data you need for server-side generated tasks and attaching them to the task (or whatever solution we come up with, this is still being discussed AFAIK), as well as all the other stuff required to get this much of the supervisor view done
 - Idea 2: somehow hard-code a dodgy change into the view that determines what documents you can see so that supervisors can also see this exact style of document (eg write the parents into `fields.peopleWhatCanAlsoSeeThis` when creating the document, and have the view use those values as well)
 - Idea 3: create a sentinel view that somehow detects these documents, clones and moves them so that supervisors are allowed to see them, and also detects changes to *those* documents to clone them _back_ if they change

Idea 1 is weeks and weeks of work, ideas 2 and 3 are more achievable in a couple of weeks, but they are very short sited.

**Comment by garethbowen** [comment]
@SCdF and I just discussed the options and we're leaning towards "Idea 2" because it's relatively simple to do in the timeframe available and doesn't cause any complications with duplicate docs and so on.

**Comment by garethbowen** [comment]
Code review please @dianabarsan

**Comment by garethbowen** [comment]
How this works...

1. Create offline roles for `chp` and `supervisor`.
2. Configure `supervisor` to have a replication depth of 2 (I think - @sglangevin is this right?). This is so they can see the CHPs they manage but not the families or reports about families.
3. Create a chp and supervisor user with the correct roles and correct contacts in the hierarchy so the supervisor is managing the chp.
4. Create an Enketo form for the mRDT test which takes the photo and has a hidden field named `needs_signoff` with a truthy value (eg: `true`). This indicates reports generated with this form are to be made available to supervisors for verification.
5. As the chp submit a report.

I haven't tested but it should be possible to configure the app to
- generate tasks for the supervisor to review MRDT reports
- show verification targets for CHPs
- show verification targets for supervisors for all their CHPs
- show verification results for supervisors for each CHP on the CHPs profile page using summary cards

**Comment by SCdF** [comment]
> Configure supervisor to have a replication depth of 2 (I think - @sglangevin is this right?)

If this isn't true, I think changing `needs_signoff` from being a boolean to being a UUID that is the person who needs to do the signoff would work.

**Comment by sglangevin** [comment]
Actually, it should be 1: they only see a list of CHW areas.

**Comment by sglangevin** [comment]
I think if it's 2, they'd see the CHW person contacts and the list of families but no people.

**Comment by garethbowen** [comment]
@SCdF I made it a boolean so you can have a multiple supervisor scenario, and also change supervisors and so another can be assigned a CHP and get all their reports needing signoff.

@sglangevin The depth needs to be configured so the supervisor can see the CHP which was 2 on my system but I think I've configured my hierarchy incorrectly. Before starting AT on this make sure the test system has a correct supervisor/chw hierarchy and the depth set correctly. This feature should not require any changes to existing hierarchies or depths.

**Comment by garethbowen** [comment]
Merged and forward ported to `master`

**Comment by sglangevin** [comment]
At first glance, this is looking good. I need to do some more testing tomorrow but I was able to get the config set up easily and the supervisor can see the report as expected. I did need replication depth of 2, which seems to function slightly differently than it used to, but in a good way 😀

**Comment by garethbowen** [comment]
There's no simple way to pull in the patient's name. We could fall back to using the sender's name but I worry that could lead to confusion about that title because sometimes it would have the sender and sometimes the subject. This would also affect all other users, though it's much less likely anyone else will have unknown subjects...

**Comment by sglangevin** [comment]
Yeah, that was my worry. Though falling back on the sender name seems helpful, I think it would be confusing to have different names there depending on who is viewing the report. That said, imagine being a supervisor and seeing a long list of reports with "Unknown subject" listed. Perhaps we should change the translation to "Unknown patient"? Ideally we'd be able to show the patient name across all devices, though I get why it wouldn't be easy to do here. Maybe a small updated to the translation would suffice for testing this with users?

**Comment by garethbowen** [comment]
> Perhaps we should change the translation to "Unknown patient"?

Not all reports are about patients so this would be incorrect sometimes.

**Comment by sglangevin** [comment]
True. Hmmm. I think it will be really confusing to managers and don't see any easy solutions. I expect this will come up again in the future too.

**Comment by garethbowen** [comment]
We could show the UUID  ;)

**Comment by sglangevin** [comment]
Is there any way to further indicate that this report is there for review purposes? And if so, can we have different info in the subject line for those cases?

**Comment by garethbowen** [comment]
Ok... what about, if the subject is unknown _and_ it needs signoff then we show "Submitted by <submitter>"?

**Comment by sglangevin** [comment]
So the subject line could say "Submitted by [submitter name]"?

**Comment by garethbowen** [comment]
Sorry, GH sanitized out my angled brackets so it wasn't obvious (edited now). Yes that's what I was thinking.

**Comment by sglangevin** [comment]
Ok - talked it over with @amandacilek and we are good with that as a solution for now and will aim to get user feedback so that we can consider ways to improve in the future.

**Comment by sglangevin** [comment]
@garethbowen sorry, I realized that I never put this one back into Active Work so that you could make the change we discussed here. Moving it now.

**Comment by garethbowen** [comment]
@sglangevin It wasn't possible to check the `needs_signoff` field without additional performance implications so I've modified the default behaviour for all reports to show "submitted by ..." when the subject is unknown but the sender is. If this is unacceptable I can add the additional information needed to the view and only show "submitted by ..." for this case.

**Comment by garethbowen** [comment]
After some discussion we've come up with another solution - if the form has a `patient_name` field we can fall back to using that. That way we can still show who the form's about without having to sync the patient doc.

**Comment by garethbowen** [comment]
Merged into `2.16.x` and `master`.

**Comment by sglangevin** [comment]
This new solution works well. Moving to ready!
